File: Python solutions/98. Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:  
    def isValidBST(self, root: TreeNode) -> bool:
        # recursive with range of low and high
        # time complexity: o(n)
        # space complexity: o(n)
        # def validate(cur, low, high):
        #     if cur == None: return True
        #     if (low != None and cur.val <= low) or \
        #        (high != None and cur.val >= high): return False
        #     return validate(cur.left, low, cur.val) and validate(cur.right, cur.val, high)
        # return validate(root, None, None)
        
        # inorder traversal
        global prev
        prev = None
        def inorder(cur):
            global prev
            if cur == None: return True
            if not inorder(cur.left): return False
            if prev != None and cur.val <= prev: return False
            prev = cur.val
            return inorder(cur.right)
        return inorder(root)

File: Java solutions/98. Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private Integer prev;
    public boolean isValidBST(TreeNode root) {
        // recursive with range as parameters
        // time complexity: o(n)
        // space complexity: o(n)
        // return validate(root, null, null);
        
        // In-order traversal
        prev = null; 
        return inorder(root);
    }
    
    private boolean inorder(TreeNode cur){
        if (cur == null) return true;
        if (!inorder(cur.left)) return false;
        if (prev != null && cur.val <= prev) return false;
        prev = cur.val;
        return inorder(cur.right);
    }
    
    public boolean validate(TreeNode cur, Integer low, Integer high){
        // Empty trees are valid BSTs
        if (cur == null) return true;
        
        // The current node's value must be between low and high
        if ((low != null && cur.val <= low) || 
            (high != null && cur.val >= high)) return false;
        
        // if both left and right are valid
        return validate(cur.left, low, cur.val) && validate(cur.right, cur.val, high);
    }
}
